controller/state.go

```diff
@@ -801,6 +800,46 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *v1
 	return &compRes, nil
 }
 
+// useDiffCache will determine if the diff should be calculated based
+// on the existing live state cache or not.
+func useDiffCache(noCache bool, manifestInfos []*apiclient.ManifestResponse, sources []v1alpha1.ApplicationSource, app *v1alpha1.Application, manifestRevisions []string, statusRefreshTimeout time.Duration, log *log.Entry) bool {
+
+	if noCache {
+		log.WithField("useDiffCache", "false").Debug("noCache is true")
+		return false
+	}
+	_, refreshRequested := app.IsRefreshRequested()
+	if refreshRequested {
+		log.WithField("useDiffCache", "false").Debug("refreshRequested")
+		return false
+	}
+	if app.Status.Expired(statusRefreshTimeout) {
+		log.WithField("useDiffCache", "false").Debug("app.status.expired")
+		return false
+	}
+
+	if len(manifestInfos) != len(sources) {
+		log.WithField("useDiffCache", "false").Debug("manifestInfos len != sources len")
+		return false
+	}
+
+	revisionChanged := !reflect.DeepEqual(app.Status.GetRevisions(), manifestRevisions)
+	if revisionChanged {
+		log.WithField("useDiffCache", "false").Debug("revisionChanged")
+		return false
+	}
+
+	currentSpec := app.BuildComparedToStatus()
+	specChanged := !reflect.DeepEqual(app.Status.Sync.ComparedTo, currentSpec)
+	if specChanged {
+		log.WithField("useDiffCache", "false").Debug("specChanged")
```

Comment:
Under what conditions should I expect to see this case? I'm trying changing the app spec and syncing, but I'm not finding `specChanged` in the logs.

Reply:
This is not about sync but refreshing the app state. You need to change the app state and wait for the next refresh loop to kick in. You can't click on the "Refresh" button either as it won't reach this part of the logic.

Reply:
bah.. just realized that the controller doesn't currently allow defining the loglevel :/
pushed a new commit to enable that of the `CompareAppState` method. Just set the env var `ARGOCD_LOG_LEVEL` to `debug` and you should be able to see the log entries.

Reply:
Yep, I see plenty of `useDiffCache` messages, just not any pertaining to `specChanged`, even if I've changed the spec of the app.

Reply:
Ha, nevermind, just found it. Not sure what was wrong earlier. :-)